shop/src/main/java/ro/msg/learning/shop/service/impl/ProductServiceImpl.java

```diff
@@ -0,0 +1,104 @@
+import ro.msg.learning.shop.model.Supplier;
+import ro.msg.learning.shop.repository.ProductRepository;
+import ro.msg.learning.shop.service.ProductCategoryService;
+import ro.msg.learning.shop.service.ProductService;
+import ro.msg.learning.shop.service.SupplierService;
+import ro.msg.learning.shop.utils.Status;
+
+import java.util.Collection;
+import java.util.Optional;
+
+@Service
+@RequiredArgsConstructor
+public class ProductServiceImpl implements ProductService {
+
+    private final ProductRepository productRepository;
+    private final ProductCategoryService productCategoryService;
+    private final SupplierService supplierService;
+
+    @Override
+    public Status addProduct(Product product) {
+        Optional<Supplier> supplier=supplierService.findByName(product.getSupplier().getName());
+        if(!supplier.isPresent()){
+            return Status.SUPPLIER_NOT_FOUND;
+        }
+        Optional<ProductCategory> productCategory=productCategoryService.findByName(product.getCategory().getName());
+        if(!productCategory.isPresent()){
+            return Status.CATEGORY_NOT_FOUND;
+        }
+        product.setSupplier(supplier.get());
+        product.setCategory(productCategory.get());
+        System.out.println(product);
+        productRepository.save(product);
+        return Status.SUCCESS;
+    }
+
+    @Override
+    @Transactional
+    public Status editProduct(Integer id, Product product) {
+
+        Optional<Product> oldProduct=productRepository.findById(id);
+        if(!oldProduct.isPresent()){
+            return Status.PRODUCT_NOT_FOUND;
+        }
+        if(null!=product.getName() && !product.getName().isEmpty()){
+            oldProduct.get().setName(product.getName());
+        }
+        if(null!=product.getDescription() && !product.getDescription().isEmpty()){
+            oldProduct.get().setDescription(product.getDescription());
+        }
+        if(null!=product.getPrice()){
+            oldProduct.get().setPrice(product.getPrice());
+        }
+        if(null!=product.getWeight()){
+            oldProduct.get().setWeight(product.getWeight());
+        }
+        if(null!=product.getSupplier() && null!=product.getSupplier().getName()){
+            Optional<Supplier> supplier=supplierService.findByName(product.getSupplier().getName());
+            if(!supplier.isPresent()){
+                return Status.SUPPLIER_NOT_FOUND;
+            }
+            oldProduct.get().setSupplier(supplier.get());
+        }
+        if(null!=product.getCategory() && null!=product.getCategory().getName()){
+            Optional<ProductCategory> productCategory=productCategoryService.findByName(product.getCategory().getName());
+            if(!productCategory.isPresent()){
+                return Status.CATEGORY_NOT_FOUND;
+            }
+            oldProduct.get().setCategory(productCategory.get());
+        }
+        return Status.SUCCESS;
+    }
+
+    @Override
```

Comment:
here is another place where transactional belongs. It belongs on all operations that alter the DB.